src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,187 @@
+
+            // Check if the current node has a child for the current character
+            if (parent.child.containsKey(ch)) {
+                parent = parent.child.get(ch); // Update the current node to the child node
+
+                // If the current node represents a pattern, record its position in res
+                if (parent.pattern_ind > -1) {
+                    res.get(parent.pattern_ind).add(i);
+                }
+
+                Node output_link = parent.output_link;
+                // Follow output links to find and record positions of other patterns
+                while (output_link != null) {
+                    res.get(output_link.pattern_ind).add(i);
+                    output_link = output_link.output_link;
+                }
+            } else {
+                // If no child node exists for the character, backtrack using suffix links
+                while (parent != root && !parent.child.containsKey(ch)) {
+                    parent = parent.suffix_link;
+                }
+                if (parent.child.containsKey(ch)) {
+                    i--; // Decrement i to reprocess the same character
+                }
+            }
+        }
+        return res; // Return the positions of patterns found in the text
+    }
+
+    // Returns the count of occurrences of each pattern in the text
+    public ArrayList<Integer> getRepeatCountOfWords() {
+        ArrayList<Integer> countOfWords = new ArrayList<>();
+        for (int i = 0; i < res.size(); i++) {
+            countOfWords.add(res.get(i).size());
+        }
+        return countOfWords;
+    }
+
+    public static void main(String[] args) {
+        String[] patterns = {"ACC", "ATC", "CAT", "GCG", "C", "T"};
+        String text = "GCATCG";
+
+        AhoCorasick obj = new AhoCorasick();
+        obj.buildTrie(patterns);
+        obj.buildSuffixAndOutputLinks();
+
+        ArrayList<ArrayList<Integer>> res = obj.search(text);
+        ArrayList<Integer> countOfWords = obj.getRepeatCountOfWords();
+
+        System.out.println("Using Zero Based Indexing");
+        System.out.println("Dictonary is : ");
+        for (int i = 0; i < patterns.length; i++) {
+            System.out.println(i + ". " + patterns[i]);
+        }
+        System.out.println();
+        System.out.println("Given text is : " + text);
+        System.out.println();
+        System.out.println("-1 represents word is not in the given string");
+        for (int i = 0; i < patterns.length; i++) {
+            System.out.print(patterns[i] + " appeared " + countOfWords.get(i) + " times at indices : ");
+            if (res.get(i).isEmpty()) {
+                System.out.print(-1 + " ");
+            } else {
+                for (int endpoint : res.get(i)) {
+                    System.out.print((endpoint - patterns[i].length() + 1) + " ");
+                }
+            }
+            System.out.println();
+        }
+    }

```

Comment:
Please remove the `main` method and instead add some proper tests. Good examples to look at:
- [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
- [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

```suggestion
```

Reply:
Yes, I understand.
Thank You for your review @vil02. 
I am working on adding test-file, I will get back to you as soon as I get it done.

Reply:
> Please remove the `main` method and instead add some proper tests. Good examples to look at:
> 
> * [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
> * [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

Hey @vil02,

I have resolved the issues you have mentioned, that is :

- Added a test-file for the code.
- Removed the main method.

Please review the code. If there's anything else that needs to be resolved regarding this, please do let me know.

Thank you!!